docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,325 @@
+# Formatted ORAS CLI output
+
+## Table of Contents
+
+- [Background](#background)
+- [Guidance](#guidance)
+- [Scenarios](#scenarios)
+  - [Scripting](#scripting)
+  - [CI/CD](#cicd)
+- [Proposal and desired user experience](#proposal-and-desired-user-experience)
+  - [oras pull](#oras-pull)
+  - [oras push](#oras-push)
+  - [oras attach](#oras-attach)
+  - [oras discover](#oras-discover)
+- [FAQ](#faq)
+
+## Background
+
+ORAS has prettified output designed for humans. However, for machine processing, especially in automation scenarios, like scripting and CI/CD pipelines, developers want to perform batch operations and chain different commands with ORAS, as well as filtering, modifying, and sorting objects based on the outputs that are emitted by the ORAS command. Developers expect that ORAS output can be emitted as machine-readable text instead of only the prettified or tabular data, so that it can be used to perform other operations.
+
+The formatted output is not intended to supersede the prettified human-readable and friendly output text of ORAS CLI. It aims to provide a programming-friendly experience for developers who want to automate their workflows without needing to parse unstructured text.
+
+## Guidance
+
+Provide two major options to enable users to define the output format of ORAS commands:
+
+- Use `--output` to output a file or directory in the filesystem
+- Use `--format` to format the output of ORAS commands into different data formats or enable process output data using the given Go template
+  - Use `--format json|tree|table` to print the output in prettified JSON, tree view, or table view
+  - Use `--format '{{ GO_TEMPLATE_FUNCTION }}'` to enable extract or compose the output data using Go template functions 
+
+> [!NOTE]
+> - `--output -` and `--format` can not be used at the same time due to conflicts.
+
+See example use cases of formatted output for `oras manifest fetch`:
+
+- If user doesn't specify `--format` flag, the default output should be raw JSON data. This is the current behavior of `oras manifest fetch`.
+
+- Use `--output <file>` and `--format` at the same time, a manifest file should be produced in the filesystem and the `mediaType` value should be outputted on the console:
+
+```console
+$ oras manifest fetch $REGISTRY/$REPO:$TAG --output sample-manifest.json --format {{.config.MediaType}}
```

Comment:
Suggest to remove `.json` suffix to avoid confusion, user might think `.json` some how asserts the output type of the file but it doesn't.
Also `.MediaType` should be lower-cased.
```suggestion
$ oras manifest fetch $REGISTRY/$REPO:$TAG --output sample-manifest --format {{.config.mediaType}}
```